ceph_deploy/osd.py

```diff
@@ -388,6 +388,228 @@ def activate(args, cfg):
         distro.conn.exit()
 
 
+def destroy(args):
+    cluster = args.cluster
+
+    for hostname, disk, journal in args.disk:
+        if not hostname:
+            raise RuntimeError('invalided hostname')
+        LOG.debug(
+            'Destroy osd id %s in host %s',
+            args.osd_id, hostname
+        )
+
+        distro = hosts.get(hostname, username=args.username)
+        LOG.info(
+            'Distro info: %s %s %s',
+            distro.name,
+            distro.release,
+            distro.codename
+        )
+
+        destroy_osd(distro, cluster, args.osd_id)
+
+
+def destroy_osd(distro, cluster, osd_id):
+
+    found_in_node = False
+    found_in_stray = False
+    cluster_path = paths.osd.base(cluster)
+    conn = distro.conn
+
+    if not osd_id or \
+       osd_id is None:
+        LOG.info("(NOT IMPLEMENT) remove all osd")
+        raise NotImplementedError("It need to assign the specific osd id.")
+
+    command = [
+        'ceph',
+        '--cluster={cluster}'.format(cluster=cluster),
+        'osd',
+        'tree',
+        '--format=json',
+    ]
```

Comment:
After looking at this for a bit, I have a few suggestions/requests.  Starting here, I don't think we need to call ceph osd tree explicity here.  Contained in this osd.py source file there is already an osd_tree() function that can be re-used.

I also think it would be worthwhile to look at the code in osd_list().  Specifically the part about gathering the osd tree from a monitor node.  This code, as written, goes to each server hosting the OSD and calls "ceph osd tree".  however, it is possible that OSD nodes do not have the cephx admin key that allows that command to work.  It would be better to gather the osd tree once from a monitor node up in destroy(), before going into the loop.